src/lib/BitcoinTxnParser.sol

```diff
@@ -0,0 +1,251 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.28;
+
+/// @title BitcoinTxnParser
+/// @notice Library for parsing Bitcoin transactions and extracting OP_RETURN metadata
+/// @dev Handles both legacy and segwit transaction formats
+library BitcoinTxnParser {
+    // Custom errors
+    error INVALID_OP_RETURN_OUTPUT();
+    error UNSUPPORTED_PUSH_OPERATION();
+    error INVALID_PUSH_DATA_1();
+    error INVALID_PUSH_DATA_2();
+    error INVALID_OP_RETURN_DATA_LENGTH();
+    error NO_OP_RETURN_FOUND();
+    error INVALID_TRANSACTION_FORMAT();
+    error INVALID_VAR_INT_FORMAT();
+
+    struct Output {
+        uint64 value;
+        bytes script;
+    }
+
+    struct Transaction {
+        uint32 version;
+        Output[] outputs;
+        uint32 locktime;
+    }
+
+    struct TransactionMetadata {
+        address receiverAddress;
+        uint256 lockedAmount;
+        uint32 chainId;
+        uint256 baseTokenAmount;
+    }
+
+    /// @notice Decodes metadata from OP_RETURN data into a structured format
+    /// @param data The raw metadata bytes from OP_RETURN output
+    /// @return Structured metadata containing receiver address, locked Amount, chain ID, and base token amount
+    /// @dev Uses assembly for efficient byte manipulation and data extraction
+    function decodeMetadata(bytes memory data) internal pure returns (TransactionMetadata memory) {
+        if (data.length != 48) {
+            // Subject to change based on metadata format
+            revert INVALID_OP_RETURN_DATA_LENGTH();
+        }
+        address receiverAddress;
+        uint256 lockedAmount;
+        uint32 chainId;
+        uint256 baseTokenAmount;
+
+        assembly {
```

Comment:
can you specify in the code in what manner data was encoded, which makes the below code readable. IMO you don't need the yul to read that encoded data.